File: llm_generated_code/ChatGPT/CPP/Level1/Q11.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(vector<vector<int>>& grid, int row, int col) {
    int m = grid.size();
    int n = grid[0].size();

    // Boundary check and check if it's land (1)
    if (row < 0 || row >= m || col < 0 || col >= n || grid[row][col] != 1) {
        return;
    }

    // Mark this cell as visited (changing 1 to 2)
    grid[row][col] = 2;

    // Explore in all four directions
    dfs(grid, row - 1, col);
    dfs(grid, row + 1, col);
    dfs(grid, row, col - 1);
    dfs(grid, row, col + 1);
}

int numEnclosedLandCells(vector<vector<int>>& grid) {
    int m = grid.size();
    int n = grid[0].size();

    // Step 1: Perform DFS on all boundary cells that are land (1)
    for (int i = 0; i < m; i++) {
        // Left boundary
        if (grid[i][0] == 1) {
            dfs(grid, i, 0);
        }
        // Right boundary
        if (grid[i][n - 1] == 1) {
            dfs(grid, i, n - 1);
        }
    }

    for (int j = 0; j < n; j++) {
        // Top boundary
        if (grid[0][j] == 1) {
            dfs(grid, 0, j);
        }
        // Bottom boundary
        if (grid[m - 1][j] == 1) {
            dfs(grid, m - 1, j);
        }
    }

    // Step 2: Count all remaining land cells (1s) that are not marked
    int enclosedLandCells = 0;
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (grid[i][j] == 1) {
                enclosedLandCells++;
            }
        }
    }

    return enclosedLandCells;
}

int main() {
    vector<vector<int>> grid = {
        {0, 0, 1, 0, 0},
        {0, 1, 1, 1, 0},
        {0, 1, 0, 1, 0},
        {0, 1, 1, 1, 0},
        {0, 0, 1, 0, 0}
    };

    int result = numEnclosedLandCells(grid);
    cout << "Number of enclosed land cells: " << result << endl;

    return 0;
}

File: llm_generated_code/ChatGPT/Python/Level1/Q11.py
def num_enclosed_land_cells(grid):
    rows, cols = len(grid), len(grid[0])
    
    def dfs(r, c):
        # If out of bounds or sea cell, stop recursion
        if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] != 1:
            return
        # Mark the cell as visited by setting it to a different value, e.g., -1
        grid[r][c] = -1
        
        # Explore in all four directions
        dfs(r + 1, c)
        dfs(r - 1, c)
        dfs(r, c + 1)
        dfs(r, c - 1)
    
    # Run DFS on all boundary land cells
    for i in range(rows):
        if grid[i][0] == 1:  # Left boundary
            dfs(i, 0)
        if grid[i][cols - 1] == 1:  # Right boundary
            dfs(i, cols - 1)
    for j in range(cols):
        if grid[0][j] == 1:  # Top boundary
            dfs(0, j)
        if grid[rows - 1][j] == 1:  # Bottom boundary
            dfs(rows - 1, j)
    
    # Count the remaining unvisited land cells
    enclosed_land_count = 0
    for i in range(rows):
        for j in range(cols):
            if grid[i][j] == 1:
                enclosed_land_count += 1
    
    return enclosed_land_count

